### PR TITLE
[BBT#237] Allow sender to send data of any size.

### PR DESCRIPTION
**Pull request by bosilca**
https://bitbucket.org/icldistcomp/parsec/pull-requests/237

The sender can send less data than expected by the receiver. If we select the communication protocol based on how much data the receiver expects, the sender and receiver could diverge on the protocol to be used (mainly visible when the sent message is below eager, while the receiver expects more data. So, force the sender to pack the amount of data per dep, and force the receiver to abide by this amount.

### CODE REVIEW

File: parsec/remote_dep_mpi.c [hunk 1]
```diff
@@ -929,13 +927,14 @@ remote_dep_get_datatypes(parsec_execution_stream_t* es,
             origin->msg.task_class_id = dtd_task->super.task_class->task_class_id;
```
Comment (abouteiller) [inline parsec/remote_dep_mpi.c:927]: In the debug printout above we still use local_mask. I don't really understand why we now use `(1U<<k)` on line 930, but the print should be consistent with that change

Comment (bosilca) [reply parsec/remote_dep_mpi.c:927]: This code has been copied from the PTG, where we do have the opportunity to handle multiple deps in same time, which is exactly what the mask identifies. However, in DTD there is no such thing as being able to call iterate_successors with a mask that has more than one active bits. So, while the local_mask contains (1U<<k), I think it is more clear to not use it in iterate_successors, but instead highlight that we call it with a single enabled bit.

**Review by devreal — COMMENTED** [review]
LGTM, although I am not sure about the details of the remote dependencies...

**Review by devreal — APPROVED** [review]

File: parsec/remote_dep_mpi.c [hunk 2]
```diff
@@ -907,13 +905,13 @@ remote_dep_get_datatypes(parsec_execution_stream_t* es,
...
                  * we later try to activate a task for which we have already received
                  * an activation for) we do not need to store the buffer and we send
-                 * PARSEC_DTD_SKIP_SAVING as an indicaton of that.
+                 * PARSEC_DTD_SKIP_SAVING as an indicator of that.
                  */
                 if( storage_id != PARSEC_DTD_SKIP_SAVING) {
                     char* packed_buffer;
                     /* Copy the short data to some temp storage */
                     packed_buffer = malloc(origin->msg.length);
-                    memcpy(packed_buffer, origin->eager_msg + *position, origin->msg.length);
```
[the remark this answers is not in the document]
Comment (therault) [reply parsec/remote_dep_mpi.c:888]: It's right if and only if `*position == 0`... Which is hard to verify by just looking at the code. Maybe we could add an 'assert(0 == *position);` above?

**Review by therault — APPROVED** [review]
Looks good to me (and my limited understanding of the comm engine).

File: parsec/remote_dep_mpi.c [hunk 3]
```diff
@@ -876,14 +873,17 @@ remote_dep_get_datatypes(parsec_execution_stream_t* es,
...
+                                               (1U<<k),
                                                remote_dep_mpi_retrieve_datatype,
                                                origin);
         }
     } else {
         parsec_task_t task;
         task.taskpool   = origin->taskpool;
-        task.priority   = 0;  /* unknown yet */
+        int idx, *data_sizes = (int*)origin->eager_msg;
+        /* Do not set the task.task_class here, because it might trigger a race condition in DTD */
```
Comment (abouteiller) [inline parsec/remote_dep_mpi.c:884]: comment says we should not set the task class, but then we do it 2 lines below?

File: parsec/remote_dep_mpi.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (abouteiller) [reply parsec/remote_dep_mpi.c:863]: I do not understand what guarantees that position == 0 is true in the dtd case

[the remark this answers is not in the document]
Comment (abouteiller) [reply parsec/remote_dep_mpi.c:863]: I put the assert and it triggers so it is not true (redistribute, redistribute_random, pingpong ctests)

[the remark this answers is not in the document]
Comment (abouteiller) [reply parsec/remote_dep_mpi.c:863]: I double checked and we enter with position != 0 already accumulated in eager_msg, the position is skipping over the header, so the code is correct. I added a comment to clarify

File: parsec/remote_dep_mpi.c (hunk 3 above)
Comment (abouteiller) [reply parsec/remote_dep_mpi.c]: I removed the comment it works otherwise

**Comment by abouteiller** [comment]
> I need to take another look at this, and solve the conflicts.

I solved the conflicts a couple of days ago and merged it in your branch please double check that it all LGTY and merge then

**Comment by abouteiller** [comment]
Saw the following error in the CI, retrying to see if it is a CI problem
```
111/116 Test #111: profiling/bw_check_hdf5 .........................................***Failed    0.31 sec
Unexpected sum of length for events of type MPI_ACTIVATE. Expected {'nb': 100, 'lensum': 11200}, found 12000.
HDF5 file bw.h5 does not contains the events or event_types attribute. The file might be empty, corrupted or it was incorrectly generated
Found 100 events of type MPI_ACTIVATE -- correct
```

**Comment by bosilca** [comment]
Let's review this again. I had some issues with `git pull` between @abouteiller changes and my own, and I hope I solve them all correctly (even if the order of the commits seems now to be different). Anyway, it passes all the tests so Yay!

**Comment by bosilca** [comment]
The profiling error is due to a hand-written value of the message size for the `MPI_ACTIVATE` messages. I think the following patch should fix it, but @therault might want to confirm.

```diff
diff --git a/tests/profiling/check-comms.py b/tests/profiling/check-comms.py
index cafc0edf3..8a82e8238 100644
--- a/tests/profiling/check-comms.py
+++ b/tests/profiling/check-comms.py
@@ -7,7 +7,7 @@ assert os.path.isfile(filename)
 t = pd.HDFStore(filename)
 
 result = {
-    'MPI_ACTIVATE': { 'nb': 100, 'lensum': 11200 },
+    'MPI_ACTIVATE': { 'nb': 100, 'lensum': 12000 },
     'MPI_DATA_CTL': { 'nb': 100, 'lensum': 209715200 },
     'MPI_DATA_PLD_SND': { 'nb': 100, 'lensum': 209715200 },
     'MPI_DATA_PLD_RCV': { 'nb': 100, 'lensum': 209715200 }
```

The GPU error seems genuine. I can't replicate for some reason, and I am unable to connect to the container running the tests.

**Comment by QingleiCao** [comment]
I will test this PR using HiCMA_PaRSEC.

**Comment by QingleiCao** [comment]
This PR works fine for small tests on 4 nodes in HiCMA_PaRSEC with `PARSEC_DIST_SHORT_LIMIT` ON. 

But when `PARSEC_DIST_SHORT_LIMIT` is OFF, it shows:
```
x@00000 The header plus data cannot be sent on a single message (need 112 but have 116)
 @remote_dep_mpi_pack_dep:1292  (a05:27295)
```

**Comment by abouteiller** [comment]
> This PR works fine for small tests on 4 nodes in HiCMA_PaRSEC with `PARSEC_DIST_SHORT_LIMIT` ON.
> 
> But when `PARSEC_DIST_SHORT_LIMIT` is OFF, it shows:
> 
> ```
> x@00000 The header plus data cannot be sent on a single message (need 112 but have 116)
>  @remote_dep_mpi_pack_dep:1292  (a05:27295)
> ```

DIST_SHORT_LIMIT is an integer (in KB), did you set it to 0 (rather than 1, default value)?

**Comment by QingleiCao** [comment]
> > This PR works fine for small tests on 4 nodes in HiCMA_PaRSEC with `PARSEC_DIST_SHORT_LIMIT` ON.
> > But when `PARSEC_DIST_SHORT_LIMIT` is OFF, it shows:
> > ```
> > x@00000 The header plus data cannot be sent on a single message (need 112 but have 116)
> >  @remote_dep_mpi_pack_dep:1292  (a05:27295)
> > ```
> 
> DIST_SHORT_LIMIT is an integer (in KB), did you set it to 0 (rather than 1, default value)?
> 
> I can replicate by setting it to 0 on the reshape test (among others), investigating.

`PARSEC_DIST_SHORT_LIMIT` is set to 1 by default if ON and 0 if OFF.

**Comment by abouteiller** [comment]
@bosilca  please check my latest commit: I reserved MAX_PARAM_COUNT+1 space for data_sizes entries, but I certainly hope not, but was wondering if we have to reserve (MAX_PARAM_COUNT+1)*MAX_DEP_COUNT as that would be several KB of waste.

**Comment by bosilca** [comment]
I think technically this should be `MAX_DEP_IN_COUNT`, as this is the maximum number of output dependencies for a task.

**Review by abouteiller — APPROVED** [review]

**Comment by bosilca** [comment]
@QingleiCao cna you please confirm that this new version passes all hicma tests?

**Comment by QingleiCao** [comment]
> @QingleiCao can you please confirm that this new version passes all hicma tests?

Yeah, this new version passes all tests in HiCMA on a small scale. 
BTW, is there any guide using this PR for bin packing in arena?
@bosilca

**Review by QingleiCao — APPROVED** [review]

**Pull request merged**